src/context/user_context.js

```diff
@@ -1,15 +1,32 @@
-import React, { useContext, useEffect, useState } from 'react';
+import React, { useContext, useReducer, useEffect, useState } from 'react';
 import axios, { AxiosError } from 'axios';
+import {
+  CURRENT_USER_FETCHED,
+  LOGIN_BEGIN,
+  LOGIN_ERROR,
+  LOGIN_SUCCESS,
+  LOGOUT_BEGIN,
+  LOGOUT_ERROR,
+  LOGOUT_SUCCESS,
+  REGISTER_BEGIN,
+  REGISTER_ERROR,
+  REGISTER_SUCCESS,
+} from '../actions';
+import auth_reducer from '../reducers/user_reducer';
 
 // Since this rootURL is used throughout the app, it might make sense to instead have a file for "Constants" that can be imported wherever needed
 const rootUrl = 'https://ecommerce-6kwa.onrender.com';
 
+const initialState = {
+  authLoading: false,
+  authError: null,
+  authState: null,
+};
+
 const UserContext = React.createContext();
 
 export const UserProvider = ({ children }) => {
-  const [currentUser, setCurrentUser] = useState(null);
-  const [authLoading, setAuthLoading] = useState(false);
-  const [authError, setAuthError] = useState(false);
+  const [authState, dispatch] = useReducer(auth_reducer, initialState);
```

Comment:
`useReducer` takes a reducer function, and an initial state. 

It returns the current state, and a `dispatch` function which we can use to trigger events/actions.